prow/tide/tide.go

```diff
@@ -866,46 +865,105 @@ func (c *changedFilesAgent) prune() {
 	c.nextChangeCache = make(map[changeCacheKey][]string)
 }
 
-func (c *Controller) presubmitsByPull(sp *subpool) (map[int]sets.String, error) {
-	presubmits := make(map[int]sets.String, len(sp.prs))
-	record := func(num int, context string) {
-		if jobs, ok := presubmits[num]; ok {
-			jobs.Insert(context)
-		} else {
-			presubmits[num] = sets.NewString(context)
+// prJobRequirements describes prowjob requirements for a PR.
+type prJobRequirements struct {
+	// requiredPJContexts are the prowjob associated status contexts that Tide
+	// expects to be green before merge.
+	requiredPJContexts sets.String
+	// allowedPendingContexts are the contexts that are allowed to be pending without
+	// removing the PR from the pool. The PR is not mergeable until these turn green,
+	// but we don't want to thrash the pool when retesting.
+	// This set is a superset of requirePJContexts that includes the contexts of
+	// any parent jobs.
+	allowedPendingContexts sets.String
+	// toTrigger is the set of presubmit jobs to trigger for this PR if we decide
+	// to retest it. This is keyed by job name instead of context since the context
+	// may be unspecified if we are triggering a parent job that doesn't report.
+	toTrigger map[string]config.Presubmit
+}
+
+func (c *Controller) requirementsByPull(sp *subpool) (map[int]prJobRequirements, error) {
+	reqs := make(map[int]prJobRequirements, len(sp.prs))
+	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
+		// Does this job require a context or have kids that might need one?
+		if !ps.AlwaysRun && ps.RunIfChanged == "" {
+			continue // No
```

Comment:
The children can't be required if the optional parent doesn't always run or have a run if changed regexp that matches the PR? Otherwise the parent wouldn't necessarily run so the child wouldn't necessarily run and can't be required.

Reply:
If there's a required child we should schedule the parent. Whether the parent would be scheduled in a world where it didn't have children should be irrelevant.

Reply:
Required children with optional parents should be invalidated IMO.